### ISSUE TITLE
End point (TAG start) marker position destroyed when in Cuesheet mode

### ISSUE THREAD

**Issue by sharmanr** [mpogue2/SquareDesk#901]
I hadn't used SquareDesk for a month or two, but have a dance tomorrow and while practising found that the end point of a singing call looked correct when MusicPlayer tab visible but not when Cuesheet was.   (See two pictures below.)

So I Unlocked For Editing, set it correctly, it displayed ok, but when I next loaded that song the problem reoccured.  It looked ok when in MusicPlay but was wrong when I actually needed it, when Cuesheet was visible.  (Because the end position is wrong the indication of what figure or break is incorrect.)

That was using the version built 13 May.   I did a "git pull" to get most recent version and found

-  "ERROR: Could not find Vamp executable. Snapping is disabled" everytime loaded song or changed tab
- Since I couldn't quickly see how to do all that was required to add the Vamp stuff I put a return before the error message (which preceded a return anyway)
- The origin problem remains :  you can set the end position but play another song and get back to it and it's gone.
- There is now another problem,  after a change when loading it now looks wrong even in when the Music Tab is visible.

Here you can see the position looks right when the MusicTab is visible but wrong (it's very near the end) when the cuesheet is visible:
<img width="1148" alt="music-player" src="https://github.com/mpogue2/SquareDesk/assets/76225465/63f42b75-59d8-4300-a66a-fe1fb642ca10">
<img width="1148" alt="cue-sheet" src="https://github.com/mpogue2/SquareDesk/assets/76225465/28b45bb0-e9f5-4aea-8fae-5060697aa09b">

I'm surprised I didn't see this behaviour before, and just realized I "upgrade" to macos 13.5 yesterday;  I wonder if there's a connection?

**Comment by sharmanr** [comment]
In the last  paragraph I meant "upgrade**d**" of course.   I doubt that an OS change would cause this behaviour.

**Comment by sharmanr** [comment]
Changing the out-position seems to be saved in the database but when a song is loaded on_dateTimeEditIntroTime_timeChanged overwrites the out-position (setting it almost to the end) and saves this new valuein the database.    What is on_dateTimeEditIntroTime_timeChanged for?  Is it meant to guess new values for patter?  Should it overwrite user-supplied values?

**Comment by sharmanr** [comment]
I made on_dateTimeEditIntroTime and on_dateTimeEditOutroTime  in mainwidow.cpp return immediately and it fixed (masked?) the problem for me

**Comment by danlyke** [comment]
I suspect, without looking at the code, that those methods are supposed to
be called by the control when the start or end times are edited.

I further suspect that something in a recent Qt has changed behavior such
that those actions get called when the value is changed from the code as
well, which means we need some sort of flag when we set the values from
code so that we don't react to that change.

**Comment by mpogue2** [comment]
See #899.  When I switched to 6.5.2, this happened, and it sounds like exactly the same problem. Switching back to 6.4.3 makes it work correctly again.  @sharmanr Which version of Qt are you testing on?

**Comment by mpogue2** [comment]
@danlyke  your analysis is correct. It is getting called too often, and we can't comment out that code....

**Comment by mpogue2** [comment]
@sharmanr  also, the Vamp warning was supposed to happen only once, whereupon it should disable the snapping function. You should not be seeing this repeatedly.  I have not checked in the vamp code yet but I can provide the binaries, if you would like...

**Comment by sharmanr** [comment]
I’m on Qt 6.6.0 — it was the latest stable when I decided to make both machines the same a while back.

**Comment by sharmanr** [comment]
Thanks, but for now I just exit that routine early.  I have a version that works so don’t want to make any changes at the moment!

**Comment by mpogue2** [comment]
My guess is that setting or modifying the end loop point will not work, with the change that you've made.
It looks to me like the problem was introduced in Qt6.5 and is still there in Qt6.6 .  

It might be something we can work around or resolve in our code base, I just don't have time to look at it right now in detail.  I'll look at this the next chance I get, because I'd like to move to Qt6.5 if we can (I don't think that Qt6.6 is out officially yet, is that right?).

**Comment by sharmanr** [comment]
Hi Milke,

Well,  don’t tell the code that,  because it did work!  I can finally get the end point right now!   Mind you,  I was working with the 2-month old code since I didn’t want to spend the time to get the new stuff going.    

I’m wondering if the recent change of the OS is causing the problem;  this sounds odd but if a change in Qt invoked similar symptoms it might be the case.

What I might to after tomorrow, when things are no longer critical, is undo my kludge,  check the problem happens again,  then try and re-install the previous OS on the laptop.  If the problem disappears then the OS is indeed a factor.   Until then I’m leaving sleeping dogs lie.   (And just hope no-one wakes them up!)

Richard

**Comment by mpogue2** [comment]
@sharmanr I wouldn't bother (but it's up to you).  I am positive that it was the Qt6.4.3 -> Qt6.5 change that uncovered the problem!  WIth Qt6.4.3, it works fine.  With Qt6.5, it fails.  I'm running Ventura 13.4.1(c) right now in both test cases, and I probably won't go to MacOS 13.5 before the weekend, but since the bug occurs on MacOS 13.4.1(c), that suggests that it was not the OS change that caused this to appear....

With your code change, try changing the end point by changing the numbers.  My guess is that it will have no effect (which would be a bug).  It's fine for the version you built, but I think I just gotta bite the bullet and debug this, to understand why Qt6.5 is working differently.  Is it a timing-related bug that is there on Qt6.4.3, but not triggered?  Could be...Unless you find the underlying root cause this week, I'll probably take a look on the weekend...

**Comment by sharmanr** [comment]
Well what bothers me is why I hadn’t noticed it before.   I haven’t updated Qt recently,  it looks like it was done on the 13th May, and I called several times after that.  So why  hadn’t I seen that the end points of songs was wrong?  Possible but strange.

**Comment by mpogue2** [comment]
For me it messes up the endpoint of the song only when I load another song.  So, the song that I just played is ok, until I load it a second time, then the endpoint of the first song is made wrong, which you only see if you load it a second time.

The simpler use case that shows this is the one in #899.  Load a song and it is ok.  Double click to load the same song again, and end loop point is now bad.

Also, when did you upgrade to 6.6? If you were on 6.4.3 before, this would have started only when you upgraded....

**Comment by mpogue2** [comment]
@sharmanr You are way ahead of me!  I just tried to move to 6.5.2 this week.  6.5.1 had display problems that made it unusable.  I would double check your end loop points, for any songs that you played since mid-May...All of mine that I played were changed (incorrectly) due to this bug.  Fortunately, I was still testing 6.5.2, so I only had a handful that needed to be re-done.

To track this down, I did a binary search to see which commit introduced this problem.  When I realized that 1.0.5 worked fine, but that ALL of the later commits that I tested had the problem, that led me to realize that 1.0.5 was compiled with 6.4.3, and all of my later-commit testing was done with 6.5.2 .  That strongly suggested that it wasn't a change that we had made, it was the Qt version that caused it.  I double checked that, by going back and forth between 6.4.3 and 6.5.2 using top-of-tree, and it was reproducible.  I am hoping to look at this in more detail this weekend... I will bet that there's a workaround.

**Comment by mpogue2** [comment]
OK, here's what is happening:

Starting in Qt6.5, the QAudioDecoder is now calling `finished()` when the decode is NOT really finished.  In fact, it has decoded literally nothing yet.  That, in turn, makes us think that the song is zero length (because there are zero frames there), which causes a bunch of divide-by-zeros which makes the Loop End Point == Infinity (this essentially gets clamped by the slider to the length of the song).  

Later on, `finished()` is called A SECOND TIME with the decode now correctly done, and sliders and dateTime widgets are set correctly, BUT the damage is already done, because the Loop End Point was already persisted from the first `finished()` call as Length-Of-Song.  Essentially, the End Loop Point information gets overwritten with a bad value by the first (incorrect) `finished()` call.

This extra call to `finished()` does NOT happen in 6.4.3, which is why we didn't see this problem there.

Code like this in `AudioDecoder::finished()` seems to work around the problem, but I want to test it some more:
```
    if ((m_input->size() == 0) && (m_data->size() == 0)) {
        qDebug() << "***** IGNORING FINISHED(), BECAUSE IT'S NOT REALLY FINISHED";
        return; // IGNORE IF IT'S NOT REALLY FINISHED
    }

```

**Comment by mpogue2** [comment]
What seems to be happening is that this line:

    `m_decoder.setSource(QUrl::fromLocalFile(fileName));`

in `AudioDecoder::setSource()` is initiating a call to `finished()`, even though the decode is already finished (song was already loaded).  This is a change in behavior from 6.4.3 .  Yes, setSource should be stopping the previous decode, BUT ONLY IF THERE WAS A DECODE IN PROGRESS.

**Comment by sharmanr** [comment]
It sounds like you found it.

I tried to build using Qt 6.3.  The Build & Run page says Qt-6.3 but compiler output shows it is using 6.6.  Annoying.

FYI: The “Could not find Vamp” happens each time select new song, when first select Cuesheet for that song, and often while moving the poisition when Cueshee it visible, and sometimes when pressing Save.

**Comment by mpogue2** [comment]
Could you kindly open a bug on the vamp dialog popping up a lot? My intent is that you would get ONE message and then it would turn off snapping so no more messages (unless you manually turned it on again).  There should be no dialog at all when moving sliders or doing a save!  Only at song load time, and then only once!  Thanks!

**Comment by mpogue2** [comment]
Fixed by: 3c41825ecf9761e242f2a5c3334336eb2bcda2b1

Starting in Qt6.5, calling setSource to setup the source filename for decoding now calls finished(), but it hasn't even started yet.  This commit detects that condition, and doesn't bother to do all the stuff that we normally do at load finished time.

**Comment by mpogue2** [comment]
See also #899.  I think this one now has a workaround, so closing it.
